internal/observability/events/context.go

```diff
@@ -0,0 +1,233 @@
+		return nil, false
+	}
+	eventer, ok := ctx.Value(eventerKey).(*Eventer)
+	return eventer, ok
+}
+
+// NewRequestInfoContext will return a context containing a value for the
+// provided RequestInfo
+func NewRequestInfoContext(ctx context.Context, info *RequestInfo) (context.Context, error) {
+	const op = "event.NewRequestInfoContext"
+	if ctx == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing context")
+	}
+	if info == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing request info")
+	}
+	if info.Id == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing request info id")
+	}
+	return context.WithValue(ctx, requestInfoKey, info), nil
+}
+
+// RequestInfoFromContext attempts to get the RequestInfo value from the context
+// provided
+func RequestInfoFromContext(ctx context.Context) (*RequestInfo, bool) {
+	const op = "event.RequestInfoFromContext"
+	if ctx == nil {
+		return nil, false
+	}
+	reqInfo, ok := ctx.Value(requestInfoKey).(*RequestInfo)
+	return reqInfo, ok
+}
+
+// WriteObservation will write an observation event.  It will first check the
+// ctx for an eventer, then try event.SysEventer() and if no eventer can be
+// found an error is returned.
+//
+// At least one and any combination of the supported options may be used:
+// WithHeader, WithDetails, WithId, WithFlush and WithRequestInfo. All other
+// options are ignored.
+func WriteObservation(ctx context.Context, caller Op, opt ...Option) error {
+	const op = "event.WriteObservation"
+	if ctx == nil {
+		return errors.New(errors.InvalidParameter, op, "missing context")
+	}
+	if caller == "" {
+		return errors.New(errors.InvalidParameter, op, "missing operation")
+	}
+	eventer, ok := EventerFromContext(ctx)
+	if !ok {
+		eventer = SysEventer()
+		if eventer == nil {
+			return errors.New(errors.InvalidParameter, op, "missing both context and system eventer")
+		}
+	}
+	opts := getOpts(opt...)
+	if opts.withDetails == nil && opts.withHeader == nil && !opts.withFlush {
+		return errors.New(errors.InvalidParameter, op, "you must specify either header or details options for an event payload")
+	}
+	if opts.withRequestInfo == nil {
+		var err error
+		if opt, err = addCtxOptions(ctx, opt...); err != nil {
+			return errors.Wrap(err, op)
+		}
+	}
+	e, err := newObservation(caller, opt...)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	if err := eventer.writeObservation(ctx, e); err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// WriteError will write an error event.  It will first check the
+// ctx for an eventer, then try event.SysEventer() and if no eventer can be
+// found an hclog.Logger will be created and used.
+//
+// The options WithId and WithRequestInfo are supported and all other options
+// are ignored.
+func WriteError(ctx context.Context, caller Op, e error, opt ...Option) {
+	const op = "event.WriteError"
+	// EventerFromContext will handle a nil ctx appropriately. If e or caller is
```

Comment:
The `If e or caller...` part of the comment should move to line 137

Reply:
the point was to put the comment here, at the top where checks for invalid parameters are typically performed so the reader doesn't have to wonder if someone just forgot to validate the inbound parameters.